### PR TITLE
Add separate QT GUI app

### PR DESCRIPTION
**Pull request by Vhuynh25**
This is not done, the device names don't work because I don't know how to format the device names from ALSA exactly for the signal_estimator.

The GUI isn't part of the signal_estimator itself, it reads the options from the ui, runs the signal_estimator separately and graphs the data it gets from a pipe. This does mean that the signal_estimator exe has to be in the same directory.

I had to make some minor changes to the signal estimator source code to do this .

- Frame.hpp now has a function that returns the iotype as an int.
- The file dumper will print to stdout if the option argument == stdout (--dump-output stdout).
- The file dumper has been changed to print out the data with the iotype prepended if and only if the option argument is stdout.

This is my first time working on open-source so please let me know if there needs to be any changes.

### CODE REVIEW

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        if (strcmp(token, "Input") == 0){
+            token = std::strtok(NULL, " \n");
+            inx.append(std::stod(token)/1000000); // append to Qvector x-axis for input graph
+            token = std::strtok(NULL, " \n");
+            if (token == NULL) // if corresponding y cannot be read, toss back the x out
+                inx.pop_back();
+            else // append to Qvector y-axis for input graph
+                iny.append(std::stod(token));
+            if (counter % 20 == 0){ // set the data for QWT plot curve in the input graph
+                incurve->setSamples(inx, iny);
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:93]: I suggest to extract parsing of signal-estimator output to a function.

Rationale: separate UI from parsing and keep boths part cleaner.

For example it can be something like:

```
struct Point {
   PointType type; // input or output
   double timestamp;
   double value;
};

std::optional<Point> parseLine(const char*);
```

Or maybe:

```
std::optional<Point> readAndParseLine(QProcess); // may be QIODevice instead of QProcess
```

File: src/Signal-Estimator-GUI/extra_options.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#include "extra_options.h"
```
Comment (gavv) [inline src/Signal-Estimator-GUI/extra_options.cpp:1]: I suggest to make two changes to file layout:

* move all existing source files from `src` to `src/cli`
* move the new UI code from `src/Signal-Estimator-GUI` to `src/gui`

I think this way it would be a bit cleaner.

Comment (gavv) [reply src/Signal-Estimator-GUI/extra_options.cpp:1]: Also, can we adjust Signal-Estimator-GUI project so that it install the resulting binary into `bin` in project root instead of `src`?

**Review by gavv — COMMENTED** [review]
Nice! Here are some comments. More to follow!

**Comment by gavv** [comment]
Regarding the UI itself. Overall, it looks very nice and plots look better than I expected - they are actually useful right now.

Below I list some minor problems.

----

### layouts

Here is how the UI looks like on my machine:

image

There are to small problems:

1. Text label are truncated.
2. The UI is not resizable and so I can't get larger plots by resizing window, but I definitely would like to see larger plots.

They both can be solved by using layouts instead of absolute positioning of widgets. Qt designer supports layouts and actually it's the recommended way to arrange interfaces. Could we do that?

-----

### share x axis

Another problem is that input and output signal plots are not aligned. Actually, ideally we should plot both input and output on the same plot which share a single X axis and vertical grid. Something similar to `sharex` option in matplotlib:



I don't remember for sure, but I hope qwt supports it.

This is important because this way you can roughly estimate the latency (the offset between input and output signal) by just looking on plots.

-----

### sliding window

Another problem with plots is that their X size constantly growths. This makes hard to estimate the signal offset because everything is moving.

So could we instead do two things?

* make the number of points on x size fixed (e.g. a few seconds, ideally it should be configurable)
* implement moving/sliding window: we always plot last N seconds; when new point arrives, we drops points before last N seconds

Ideally, we should use circular buffer for that to avoid frequent (re)allocations.

-----

### plot order

Nitpick: it would be more logical to place output signal before input signal. Because output precedes input in our case: we write output to device, it's echoed back to input device, and we capture it.

We can even add clarification in the UI like this:

```
Output (emitted) signal:

....

Input (captured) signal:

....
```

-----

### labels

Could we please also add labels to list boxes on the top? I mean input & output devices. mode, and format.

BTW, do we really need output format listbox (text/json)? How is it useful in the GUI?

-----

### extra options

It seems that options configured in "Extra options" have no effect.

**Comment by gavv** [comment]
> This is not done, the device names don't work because I don't know how to format the device names from ALSA exactly for the signal_estimator.

I think we can fix device selection later, so that now in this PR we can concentrate on the rest features.

> The GUI isn't part of the signal_estimator itself, it reads the options from the ui, runs the signal_estimator separately and graphs the data it gets from a pipe. This does mean that the signal_estimator exe has to be in the same directory.

Good idea, by the way! There is a small concern that communication via pipe can add extra latency or delays, but seems that the pipe throughput is enough for us and everything works smooth.

> This is my first time working on open-source so please let me know if there needs to be any changes.

You're welcome!

**Comment by gavv** [comment]
A final note. There are some issues with current thread communication scheme.

The current code works as follows:

* Background thread:
 
  * wait for a line from signal-estimator
  * lock mutex
  * modify QwtPlotCurve
  * unlock mutex
  * repeat

* UI thread:

  * sleep
  * lock mutex
  * tell QwtPlot to replot from QwtPlotCurve
  * tell qt to process accumulated events
  * unlock mutex
  * repeat

The problem with this approach is that while we're sleeping, and while we're waiting for a mutex on UI thread, we're blocking qt from processing other events (mouse clicks, window moves, etc), potentioally making the UI less responsive.

Such approach is considered a bad practice. The recommended way is to:

* make all event handlers (button handler in our case) return fast and without blocking
* let qt to manage event processing by itself, allowing the UI to be as responsive as possible

So, how we can fix this?

* First. instead of running a blocking loop in button handler, we can start a timer (QTimer) that invokes our callback periodically. The callback will get data from background thread and push it to the UI.
* Second, instead of accessing QwtPlotCurve from two threads and using a mutex (which may block UI thread, which is not desired), we can use a non-blocking circular buffer.

So here are an example approach that I would suggest to try:

* Get some thread-safe and lock-free circular buffer implementation, for example this one: https://github.com/jnk0le/Ring-Buffer - we can add it to 3rdparty

* Create two circular buffers: one with input points, one with output points. (It's important to use one buffer with points instead of two buffers with X and Y to ensure that X and Y are always in-sync).

* Background thread reads lines from signal-estimator, parses them, and appends points to one of the circular buffers (input or output); no locking is needed since the buffer is thread-safe itself

* UI thread starts the timer which invokes our callback periodically (e.g. every 10ms)

* The callback does the following for each of the circular buffers (input and output):

  * check how much points are available in the buffer
  * read all points from buffer (no locking is needed since the buffer is thread-safe; also this is a non-blocking operation since the buffer is lock-free)
  * appends new points to previously plotted data
  * if the result exceeds sliding window size, shift extra points from the beginning (so that we always plot only last N points)
  * pass points to QwtPlotCurve
  * ask QwtPlot to replot
  * return from callback

This way, QwtPlotCurve and QwtPlot are only accessed from the UI thread, and UI handlers are non-blocking and don't sleep or lock on mutexes.

To make the implementation cleaner, I suggest to extract the logic above to a class that looks something like this:

```
class PointsBuffer {
public:
    // called by background thread
    void appendPoint(Point);

    // called by UI thread in timer callback
    // this does everything listed above: reads new points from rinbbuf, appends to previous points,
    // truncates them to sliding window size, and returns the result
    QVector<Point> getCurrentPoints();

private:
    RingBuffer<Point> ringbuf_;
    QVector<Point> lastPlottedPoints_;
};
```

Also, to store last plotted points, instead of QVector<Point> we can use QCircularBuffer<Point>. It will allow us to avoid reallocations when appending new points because it allows to efficiently remove extra elements from the beginning.

File: src/Main.cpp [hunk 3]
```diff
@@ -100,6 +101,8 @@ int main(int argc, char** argv) {
          cxxopts::value<unsigned int>()->default_value(std::to_string(config.latency_us)))
         ("d,duration", "Measurement duration, seconds",
          cxxopts::value<float>()->default_value(std::to_string(config.duration)))
+        ("g,gui", "Open GUI",
+         cxxopts::value<bool>())
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/Main.cpp:105]: sorry I was trying different methods of implementing the gui and I must have missed some changes when I was deleting my other attempts. I think I'll start with redoing the ui with layouts first.

File: src/Signal-Estimator-GUI/main.cpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#include "mainwindow.h"
+#include <QApplication>
+#include "sigestnotfound.h"
+
+int main(int argc, char* argv[])
+{
+    QApplication a(argc, argv);
+    if (FILE* check = fopen("signal-estimator","r")){
+        fclose(check);
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/main.cpp:9]: is startSignalEstimator supposed to return a QProcess that is set up and ready to go?

[the remark this answers is not in the document]
Comment (gavv) [reply src/Signal-Estimator-GUI/main.cpp:9]: Yep, I meant that - just like popen() returns usable object.

**Comment by Vhuynh25** [comment]
Ok I made a bunch of big changes and I would like feedback on those but until then I'll work on the smaller stuff like fixing indentations.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+
+
+void MainWindow::read_graph_data(){
+     std::string buffer;
+
+     char* fullstr;
+     QByteArray arr;
+
+     this->proc->setReadChannel(QProcess::StandardOutput);
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:68]: Can we move setReadChannel() call to startSignalEstimator() ?

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        // proc emits error signal
+        this->proc->connect(this->proc,SIGNAL(error(QProcess::ProcessError)),this,SLOT(checkProc()));
+
+        QTimer* time = new QTimer(0);
+
+        this->timer = time; // timer for ui graphing
+
+        timer->setInterval(20);
+        timer->connect(timer, SIGNAL(timeout()),this,SLOT(update_graphs()));
+        this->proc->connect(this->proc,SIGNAL(readyReadStandardOutput()), this, SLOT(read_graph_data()));
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:234]: So, instead of using a blocking read in a separate thread, we can now use a non-blocking handler for readyReadStandardOutput() signal, invoked on UI thread.

That's a very good idea, I did't think of it! This makes things much simpler for us.

Since we're not using multiple threads anymore, we don't need thread-safe ring buffer (jnk0le::Ringbuffer). We can just use QCircularBuffer.

readyReadStandardOutput handler will parse points and append them to QCircularBuffer.

QTimer handler will copy current points from QCircularBuffer and plot them.

No need for second ring buffer, no need for synchronization, great!

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+        //this->proc->connect(this->proc,SIGNAL(finished(int)), this, SLOT(on_StopButton_clicked()));
+
+        if (this->proc->open(QProcess::ReadOnly) == false) // failing? to open signal-est
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:237]: Can we move open() call to startSignalEstimator? It may return nullptr when failed to open, and we'll check it here.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+
+
+void MainWindow::read_graph_data(){
+     std::string buffer;
+
+     char* fullstr;
+     QByteArray arr;
+
+     this->proc->setReadChannel(QProcess::StandardOutput);
+     arr = this->proc->readLine(); // read line from proc
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:69]: 1. We should check canReadLine() before calling readLine(). If canReadLine() returns false, than there is no complete line in QProcess yet, so we should return and wait until the signal is emitted next time.
2. It's possible that there are more than one line available for read. But the signal wont be re-emitted until *new* data arrives. So we should read lines in a loop until canReadLine() returns false.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+void MainWindow::on_StopButton_clicked()
+{
+    if (this->proc->isOpen()){this->proc->close();} // close QProcess
+
+    this->set_update_plots(false); // stop plotting when stop button is pressed
+}
+
+void MainWindow::parseLine(char* buffer){
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:257]: I suggest to move this function to ipc.h / ipc.cpp to separate parsing from UI. It may accept QString or QByteArray and return QPointF.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    // had to leave this function in mainwindow because we have to use QPointF for graphing
+    // we could have this function outside of main window but we would need to emit signals
+    // and the point buffers would need to be QObjects since connect() needs a receiver object
+    //
+    const char* token = std::strtok(buffer," \n"); // get iotype
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:262]: strtok() is not thread-safe, so it's btter to avoid it. Maybe use QRegExp instead? I think it'll make the code cleaner.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+        list.append(temp);
+    else {
+        c = (std::strstr(temp.toStdString().c_str() , "card ") + 5); // get card number
+        buffer = "hw:";
+        buffer.push_back(*c); // hw:X
+        c = (std::strstr(temp.toStdString().c_str() , " device ") + 8); // get device
+        buffer.append(","); // hw:X,
+        buffer.push_back(*c); // hw:X,Y
+        list.append(QString::fromStdString(buffer));
+    }
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:104]: Let's extract this logic into a function, say formatDeviceName(), and move the function to ipc.h / ipc.cpp

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+    QString t;
+    if ((t = this->ui->SampleRate->cleanText()) != QString::fromStdString("48000")){ // if it is not default value
+        list.append("-r");
+        list.append((t));
+    }
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:124]: Why do we need to hardcode all default values here? This code will silently break if the default changes and we forget to update it here. Can't we just always specify all options, even if they are default?

**Review by gavv — COMMENTED** [review]
Thanks for update, I've resolved discussions that were fixed and here are some more comment.

**Comment by gavv** [comment]
I've pushed a change to `main` branch that adds makefile with `fmt` target.

Could you please rebase on fresh main and run `make fmt`? It will auto-format all code under src according to `.clang-format` config.

File: src/Signal-Estimator-GUI/ipc.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    while (fgets(buffer,128,pipe2.get()) != nullptr){
+        if (std::strstr(buffer,"card") != NULL && std::strstr(buffer,"device") != NULL) // if line has both card and device in it
+            Strvec.append(QString::fromStdString(buffer)); // add to combobox
+    }
+    return Strvec;
+}
+
+QProcess* startSignalEstimator(QStringList args)
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/ipc.cpp:36]: I don't know if this is possible since the connect function for signals and slots wants raw pointers for the sender object and receiver object.

I'll see if I can make it work.

File: src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro [hunk 14]
```diff
@@ -0,0 +1,43 @@
+#-------------------------------------------------
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro]: what version of cmake are we using?

**Comment by Vhuynh25** [comment]
Can you do a quick check with signal-estimator? I'm not getting anything when using --dump-input and I need to know if it is on my end.

**Comment by Vhuynh25** [comment]
also is it normal for signal-estimator to not return to the terminal after finishing?. I have press ctrl c to exit to terminal

File: src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro (hunk 14 above)
[the remark this answers is not in the document]
Comment (gavv) [reply src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro]: >= 3.0.0

See https://github.com/gavv/signal-estimator/blob/main/CMakeLists.txt#L1

File: src/Signal-Estimator-GUI/mainwindow.h [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    Ui::MainWindow *ui;
+    bool update_plots = false;
+
+    PointsBuffer out_data;
+    PointsBuffer in_data;
+
+    QwtPlotCurve* curve1 = new QwtPlotCurve ("Input Curve");
+    QwtPlotCurve* curve2 = new QwtPlotCurve ("Output Curve");
+    QTimer* timer;
+    QPointer<QProcess> proc;
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.h:55]: QPointer does not automatically destroy object, it works like std::weak_ptr. Can we use QSharedPointer instead?

File: src/Signal-Estimator-GUI/ipc.cpp [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    QVector<QString> Strvec;
+    // run pipe to get output devices from aplay
+    std::unique_ptr<FILE,decltype(&pclose)> pipe1(popen("aplay -l","r"),pclose);
+
+    if(!pipe1){
+        return Strvec;
+    }
+
+    // read everything from aplay into result
+    while (fgets(buffer,128,pipe1.get()) != nullptr){
```
Comment (gavv) [inline src/Signal-Estimator-GUI/ipc.cpp:14]: It would be better to use sizeof(buffer) instead of 128 here to avoid unnoticed errors if we change buffer size.

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+
+
+
+void MainWindow::read_graph_data(){
+    QString buffer;
+    QPointF pt;
+    QByteArray arr;
+    while (this->proc->canReadLine()){
+        arr = this->proc->readLine(); // read line from proc
+        buffer = QTextCodec::codecForMib(106)->toUnicode(arr); // convert to std::string
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:68]: This looks a bit cryptic, can't we use just `buffer = QString(arr)` ?

File: src/Signal-Estimator-GUI/mainwindow.cpp [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+        //this->ui->ErrorLabel_2->setText(QString(arr));
+        pt = parseLine(buffer);
+        if (pt.isNull() == false && buffer.contains(QString("in"))){
+            this->in_data.appendPoint(pt);
+        }
+        else if (pt.isNull() == false && buffer.contains(QString("out"))){
```
Comment (gavv) [inline src/Signal-Estimator-GUI/mainwindow.cpp:74]: Here the string format partially "leaks" from parseLine() (i.e. it's not fully encapsulated by it), so we now have two places in code with the knowledge about the format. That's not very good, it's better to keep such things in one place. So I suggest to move "in" / "out" checks into parseLine() and return the parsed type from it - e.g. it can return a tuple with point and point type enum.

**Comment by gavv** [comment]
> Can you do a quick check with signal-estimator? I'm not getting anything when using --dump-input and I need to know if it is on my end.

I've checked out main branch and run:

```
./bin/signal-estimator -m latency -o pulse -i pulse -d 5 \
    --volume 1.0 --dump-output output.txt --dump-input input.txt
opening alsa writer for device pulse
suggested_latency: 8000 us
suggested_buffer_size: 384 samples
selected_buffer_time: 12000 us
selected_buffer_size: 576 samples
selected_period_time: 4000 us
selected_period_size: 192 samples
opening alsa reader for device pulse
suggested_latency: 8000 us
suggested_buffer_size: 384 samples
selected_buffer_time: 12000 us
selected_buffer_size: 576 samples
selected_period_time: 4000 us
selected_period_size: 192 samples
can't enable real-time scheduling policy
can't enable real-time scheduling policy
latency:  sw+hw 283.258ms  hw 276.443ms  hw_avg5 276.443ms
latency:  sw+hw 534.344ms  hw 519.462ms  hw_avg5 397.953ms
...
```

Works as expected, `input.txt` and `output.txt` have samples.

**Comment by gavv** [comment]
> also is it normal for signal-estimator to not return to the terminal after finishing?. I have press ctrl c to exit to terminal

This is not OK, but it's a known bug: https://github.com/gavv/signal-estimator/issues/5

We'll eventually fix it, but I think anyway GUI should have an additional timeout and kill signal estimator if it does not exit during timeout, because GUI should not hang even if signal-estimator hangs.

**Comment by gavv** [comment]
Please ping me when the PR will be ready for review.

**Comment by Vhuynh25** [comment]
@gavv ready for review

File: src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro (hunk 14 above)
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro]: I'm not comfortable enough with CMake to convert the gui build system. I think you should get someone else to do that.

File: src/FileDumper.cpp [hunk 19]
```diff
@@ -39,6 +39,10 @@ FileDumper::~FileDumper() {
 }
 
 bool FileDumper::open(const char* filename) {
+  if (strcmp(filename,"stdout") == 0){
+    fp_ = stdout;
+  }
+  else
     fp_ = fopen(filename, "w");
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/FileDumper.cpp]: fixed

File: src/FileDumper.cpp [hunk 20]
```diff
@@ -67,28 +71,32 @@ void FileDumper::write(Frame& frame) {
...
         = int(find_max(buf, bufsz) / config_.dump_rounding * config_.dump_rounding);
 
     const bool changed = (new_val != last_val_);
 
-    if (changed) {
+    if (changed && fp_ != stdout) {
         print_last_maybe_();
     }
+    else if (changed && iotype == 0){print_output_();}
+    else if (changed && iotype == 1){print_input_();}
```
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/FileDumper.cpp]: I kept the separate output formats so we don't have to change the python script

File: src/Signal-Estimator-GUI/ipc.cpp (hunk 16 above)
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/ipc.cpp]: done

File: src/Signal-Estimator-GUI/mainwindow.cpp (hunk 17 above)
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/mainwindow.cpp]: done

File: src/Signal-Estimator-GUI/mainwindow.cpp (hunk 18 above)
Comment (Vhuynh25) [reply src/Signal-Estimator-GUI/mainwindow.cpp]: done

File: src/FileDumper.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (gavv) [reply src/FileDumper.cpp]: > I kept the separate output formats so we don't have to change the python script

What is the rationale? I think that it's surprising and inconvenient for the user that changing dump destination would also change dump format.

File: src/FileDumper.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (gavv) [reply src/FileDumper.cpp]: Still not fixed actually, here and in few other places...

image

image

image

Maybe you could just run code formatter I've mentioned earlier? (`make fmt`)

File: src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro (hunk 14 above)
[the remark this answers is not in the document]
Comment (gavv) [reply src/Signal-Estimator-GUI/Signal-Estimator-GUI.pro]: No worries. We'll do the switch later.

File: src/Signal-Estimator-GUI/ipc.cpp (hunk 16 above)
Comment (gavv) [reply src/Signal-Estimator-GUI/ipc.cpp]: The same for getInputDevices().

File: src/gui/points.h [hunk 21]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+#include <optional>
+#include <iostream>
+#include <cstring>
+#include "3rdparty/Ring-Buffer-master/ringbuffer.hpp"
+#include <QVector>
+#include <QPointF>
+
+typedef enum {Input,Output, None} io ;
```
Comment (gavv) [inline src/gui/points.h:9]: `io` looks a bit too short for a type name, also type names are usually capitalized, so I'd suggest something like `IOType`. Also, typedef like this is not needed in C++. And we can use enum class here.

File: CMakeLists.txt [hunk 22]
```diff
@@ -71,12 +71,7 @@ target_link_libraries(signal-estimator
   asound
   m
   ${CMAKE_THREAD_LIBS_INIT}
-  )
-
-install(
-  TARGETS signal-estimator
-  RUNTIME DESTINATION bin
-  )
```
Comment (gavv) [inline CMakeLists.txt:79]: Why is this removed?

Comment (Vhuynh25) [reply CMakeLists.txt:79]: weird, I don't remember removing that. All I did was change it to work with the reorganized src directory.

File: src/FileDumper.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/FileDumper.cpp]: I was prioritizing the python script and I believed that if users weren't printing to stdout, they would choose sensible output file names to distinguish between the output dump and the input dump. If you still want to use the new dump format for all destinations, let me know so I can change it.

File: CMakeLists.txt (hunk 22 above)
Comment (Vhuynh25) [reply CMakeLists.txt]: I don't know why it was removed. I only edited the add_executable section to fit the src directory layout. I just added it to the cmakelist.

File: src/FileDumper.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/FileDumper.cpp]: I thought users would sensibly name their dump destinations to distinguish between input and output. If you want to use the new format, let me know so I can change it.

[the remark this answers is not in the document]
Comment (Vhuynh25) [reply src/FileDumper.cpp]: @gavv do you want me to change it or anything else?

[the remark this answers is not in the document]
Comment (gavv) [reply src/FileDumper.cpp]: Users may change destination here and there. Imagine you're debugging a shell pipeline that processes signal estimator output. One time you can redirect stdout to your pipeline, another time you can ask signal-estimator to save output to file and then feed the file to your pipeline. It would be inconvenient if your pipeline stops working and you need to adjust it depending on whether you're using files or pipes.

Another example. It would be surprising if these two commands will produce different results:

```
signal-estimator --dump-output "-" >file.txt
```

```
signal-estimator --dump-output file.txt
```

So yes, I suggest to unify the format,

**Comment by gavv** [comment]
Thanks for update. To sum up, most important remaining changes are:

* get rid of jnk0le::Ringbuffer dependency, use circular buffer from qt
* unify dump format
* fix formatting

**Comment by Vhuynh25** [comment]
@gavv I don't think we use the circular buffer from qt. I've looked into it and it looks like it's only available for qt 5.5. I'm using qt 5.9.5 so I installed and checked the required library "qt3d" and it wasn't there. I probably have to install qt 5.5 to use it which would mean anyone who wants to build the gui would also have to use qt 5.5. I have to ask what is wrong with using the third-party ring buffer?

**Comment by gavv** [comment]
> @gavv I don't think we use the circular buffer from qt. I've looked into it and it looks like it's only available for qt 5.5. I'm using qt 5.9.5 so I installed and checked the required library "qt3d" and it wasn't there. I probably have to install qt 5.5 to use it which would mean anyone who wants to build the gui would also have to use qt 5.5. I have to ask what is wrong with using the third-party ring buffer?

All we need now is non-threadsafe ring buffer, this is a very simple thing (compared to thread-safe & lock-free one) and I think it's an overkill to add external dependency just for this.

If there are problems with QCircularBuffer, we can just use vector + current offset in vector + `%` operator.

Also, the current implementation of PointsBuffer is strange, I hoped it to be revised after porting to QCircularBuffer. But we can revise it without QCircularBuffer too.

The problematic part is the following:

```
    while (lastPlottedPoints.count() >= ringbuf_.readAvailable()) {
        lastPlottedPoints.pop_front();
    }
    for (int index = 0; index < ringbuf_.readAvailable(); index++){
        if (!lastPlottedPoints.contains(ringbuf_[index])){
            // last plotted points does not have element from ringbuf, append
            lastPlottedPoints.append(ringbuf_[index]);
        }
    }
```

The second loop here has quadratic complexity, i.e. O(N^2) - we iterate all elements in first buffer and for each element, contains() iterates all elements in the second buffer.

But this is not needed at all. Now, when we're always working from a single thread, we can just copy buffer contents once with O(N) complexity.

Repetitive calls to pop_front() and append() are not needed as well. They are harmful because they may produce extra reallocations. But instead we can just allocate output vector once and copy current ring buffer contents to it.

So my suggestion is to rework PointsBuffer:

* don't use jnk0le::Ringbuffer and remove it from 3rdparty and cmake
* fix getCurrentPoints() to avoid unnecessary complexity

**Comment by gavv** [comment]
BTW do we really need std::sort() in MainWindow::update_graphs()? Could you explain why?

**Comment by gavv** [comment]
Also forgot one small thing. As you can see, in this project files are named in CamelCase (the file name is the same as the class name inside it). Please rename new files accordingly.

**Comment by gavv** [comment]
Thanks for update, is it ready for review & merge?

**Comment by Vhuynh25** [comment]
@gavv are any other changes you want to add?

**Comment by gavv** [comment]
> @gavv are any other changes you want to add?

Nothing except we've discussed above. I'll merge the PR and prepare a couple of follow-up commits, and create tasks for future improvements. Thanks a lot, huge work!

**Comment by gavv** [comment]
I've fixed code style and a few minor issues (some of them were discussed above, and a couple are new).

I've split changes into commits with descriptions, so that you could find what I've changed:

* c486e495a2194616566733422b67b3c782bbedf9
* 8225766848b2a2d14a423ef92193ceaf0d0ec7fc
* d30e243a495229a1b959b45eef6b004d24d05f22
* 83ad94f1d4fa20e9edca0d43e0d3c5a08841dec1
* dd8ea3e648bea1b56243af5402d2d320c840b0ab
* cb1724cb46ae1265983514330f8b65c18d2fb92c
* 4501590716f8d6b2086014ce08ebc81d1e15427a
* 8e12f157d5ccfbc7569bb018649e14deb1b6c2c4
* 5ece02e202d7d77d262dd14a9badbf8f8e5cec32
* 613801fdd836b9c03ded9500c1f051737c89de5e

**Comment by gavv** [comment]
Tasks for further GUI improvements:

* #17
* #13
* #14
* #15
* #16

**Comment by gavv** [comment]
One more follow-up commit: 8c2bc8defb1dbe60e9df574cf28910f866cbca2e
